# Khảo sát số bước lặp để kiểm tra một mảng có được sắp xếp tăng hay không.

In [3]:
import numpy as np
import random

In [23]:
# Hàm sinh mảng ngẫu nhiên và đếm số vòng lặp
def counting_loops(array_size, max_value):
    testing_arr_ = np.asarray(np.random.randint(max_value, size = array_size))

    loop_counting = 0;
    for i in range(0, array_size - 1):
        loop_counting += 1;

        # Điều kiện dừng: xuất hiện 1 cặp nghịch thế.
        if testing_arr_[i] > testing_arr_[i + 1]: break;

    return loop_counting

In [24]:
# Hàm tính kỳ vọng của một dictionary là bảng biến cố / tần suất
def expected_value(table):
    result = 0
    for variable, frequency in table.items():
        result += frequency * variable
    return result

In [26]:
# Hàm thực hiện khảo sát
def do_test(total_samples, array_size, max_value):
    expected_values_ = {}

    for i in range(0, total_samples):
        loop_counting = counting_loops(array_size, max_value)
        if loop_counting in expected_values_:
            expected_values_[loop_counting] += 1
        else:
            expected_values_[loop_counting] = 1

    for variable, frequency in expected_values_.items():
        expected_values_[variable] = frequency / total_samples
    
    return expected_values_

# Tiến hành khảo sát

## 1. Khảo sát trên mảng cặp $m = 1000, n = 1000$

In [33]:
# Khảo sát nhanh trên 10000 mảng có kích thước 1000 phần tử trong miền {0, .., 1000}
TOTAL_SAMPLES = 10000
ARRAY_SIZE = 1000 # m
MAX_VALUE = 1000  # n

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.5, 4: 0.02, 3: 0.22, 2: 0.26}
1.76


In [31]:
# Khảo sát nhanh trên 100000 mảng có kích thước 1000 phần tử trong miền {0, ..., 1000}
TOTAL_SAMPLES = 100000

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.508, 2: 0.321, 3: 0.124, 5: 0.01, 7: 0.001, 4: 0.033, 8: 0.001, 6: 0.001, 9: 0.001}
1.7339999999999998


In [32]:
# Khảo sát nhanh trên 10^6 mảng có kích thước 1000 phần tử trong miền {0, ..., 1000}
TOTAL_SAMPLES = 1000000

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.499255, 3: 0.125204, 2: 0.333637, 4: 0.033441, 5: 0.007126, 7: 0.000162, 6: 0.001154, 8: 1.8e-05, 9: 3e-06}
1.7197639999999998


## Khảo sát trên cặp mảng $m = 10000, n = 1000$

In [34]:
# Khảo sát nhanh trên 10000 mảng có kích thước 1000 phần tử trong miền {0, .., 1000}
TOTAL_SAMPLES = 10000
ARRAY_SIZE = 10000 # m
MAX_VALUE = 1000  # n

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.54, 2: 0.33, 3: 0.09, 4: 0.03, 5: 0.01}
1.6400000000000003


In [38]:
# Khảo sát nhanh trên 100000 mảng có kích thước 1000 phần tử trong miền {0, .., 1000}
TOTAL_SAMPLES = 100000
ARRAY_SIZE = 10000 # m
MAX_VALUE = 1000  # n

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.49761, 2: 0.33321, 4: 0.03406, 3: 0.12697, 5: 0.00674, 6: 0.00116, 7: 0.00022, 9: 1e-05, 8: 2e-05}
1.72363


In [39]:
# Khảo sát nhanh trên 10^6 mảng có kích thước 10000 phần tử trong miền {0, .., 1000}
TOTAL_SAMPLES = 1000000
ARRAY_SIZE = 10000 # m
MAX_VALUE = 1000  # n

expected_values_ = do_test(TOTAL_SAMPLES, ARRAY_SIZE, MAX_VALUE)

print(expected_values_)
print(expected_value(expected_values_))

{1: 0.500117, 2: 0.333006, 3: 0.125416, 4: 0.033112, 5: 0.006943, 6: 0.001201, 7: 0.000178, 8: 2.5e-05, 9: 2e-06}
1.7182100000000005


Dễ dàng nhận thấy, các giá trị $E(X)$ dần hội tụ về 1.72